### PR TITLE
discovery+switch: apply zero forwarding policy updates in-memory as w…

### PR DESCRIPTION
**Pull request by valentinewallace**
…ell as on disk

In this commit, we fix a bug where if a user updates a forwarding policy to be
zero, the update will be applied to the policy correctly on-disk, but not
in-memory.

We solve this issue by having the gossiper return the list of on-disk updated
policies and passing these policies to the switch for updates, so the switch can assume
that zero-valued fields are intentional and not just uninitialized.

### CODE REVIEW

File: discovery/gossiper.go [hunk 1]
```diff
@@ -345,20 +346,23 @@ func New(cfg Config, selfKey *btcec.PublicKey) *AuthenticatedGossiper {
 // AuthenticatedGossiper ensures the update has been committed by dependent
 // sub-systems, then it signs and broadcasts new updates to the network.
 func (d *AuthenticatedGossiper) PropagateChanPolicyUpdate(
-	newSchema routing.ChannelPolicy, chanPoints ...wire.OutPoint) error {
+	newSchema routing.ChannelPolicy, chanPoints ...wire.OutPoint) (
+	map[wire.OutPoint]*channeldb.ChannelEdgePolicy, error) {
```
Comment (wpaulino) [inline discovery/gossiper.go]: Nit: should document return parameters.

File: discovery/gossiper.go [hunk 2]
```diff
@@ -1292,10 +1303,13 @@ func (d *AuthenticatedGossiper) processChanPolicyUpdate(
 			edge: edge,
 		})
 
+		// Finally, add the edge to our policy mapping.
+		chanPolicies[info.ChannelPoint] = edge
```
Comment (wpaulino) [inline discovery/gossiper.go]: We should only be setting these for channels that have actually been updated (further below).

File: discovery/gossiper_test.go [hunk 3]
```diff
@@ -3353,13 +3353,23 @@ func TestPropagateChanPolicyUpdate(t *testing.T) {
 
 	const numChannels = 3
 	channelsToAnnounce := make([]*annBatch, 0, numChannels)
+	channelPoints := make([]OptionalMsgField, 0, numChannels)
```
Comment (wpaulino) [inline discovery/gossiper_test.go]: Alternatively, you could make unique channel points within `createAnnouncements` the same way without having to use the functional options.

File: htlcswitch/link.go [hunk 4]
```diff
@@ -2139,31 +2139,19 @@ func (l *channelLink) AttachMailBox(mailbox MailBox) {
...
-	if newPolicy.TimeLockDelta != 0 {
-		l.cfg.FwrdingPolicy.TimeLockDelta = newPolicy.TimeLockDelta
-	}
-	if newPolicy.MinHTLC != 0 {
-		l.cfg.FwrdingPolicy.MinHTLC = newPolicy.MinHTLC
-	}
+	l.cfg.FwrdingPolicy.BaseFee = newPolicy.BaseFee
+	l.cfg.FwrdingPolicy.FeeRate = newPolicy.FeeRate
+	l.cfg.FwrdingPolicy.TimeLockDelta = newPolicy.TimeLockDelta
+	l.cfg.FwrdingPolicy.MinHTLC = newPolicy.MinHTLC
```
Comment (wpaulino) [inline htlcswitch/link.go]: Looks like we can just do `l.cfg.FwrdingPolicy = newPolicy`?

File: htlcswitch/link_test.go [hunk 5]
```diff
@@ -5508,9 +5508,9 @@ func TestForwardingAsymmetricTimeLockPolicies(t *testing.T) {
 	// Now that each of the links are up, we'll modify the link from Alice
 	// -> Bob to have a greater time lock delta than that of the link of
 	// Bob -> Carol.
-	n.firstBobChannelLink.UpdateForwardingPolicy(ForwardingPolicy{
-		TimeLockDelta: 7,
-	})
+	newPolicy := n.firstBobChannelLink.cfg.FwrdingPolicy
+	newPolicy.TimeLockDelta = 7
+	n.firstBobChannelLink.UpdateForwardingPolicy(newPolicy)
```
Comment (wpaulino) [inline htlcswitch/link_test.go]: Why is this change needed?

File: htlcswitch/switch.go [hunk 6]
```diff
@@ -445,55 +445,63 @@ func (s *Switch) SendHTLC(firstHop lnwire.ShortChannelID, paymentID uint64,
...
 	}
 
 	return nil
 }
 
+// dbPolicyToFwdingPolicy is a helper function that converts a channeldb
+// ChannelEdgePolicy into a ForwardingPolicy struct for the purpose of updating
+// the forwarding policy of a link.
+func dbPolicyToFwdingPolicy(policy *channeldb.ChannelEdgePolicy) ForwardingPolicy {
+	return ForwardingPolicy{
```
Comment (wpaulino) [inline htlcswitch/switch.go]: Missing `MaxHTLC` field?

File: htlcswitch/switch.go [hunk 7]
```diff
@@ -445,55 +445,63 @@ func (s *Switch) SendHTLC(firstHop lnwire.ShortChannelID, paymentID uint64,
...
 // forwarding policies for all links have been updated, or the switch shuts
 // down.
-func (s *Switch) UpdateForwardingPolicies(newPolicy ForwardingPolicy,
-	targetChans ...wire.OutPoint) error {
-
-	log.Debugf("Updating link policies: %v", newLogClosure(func() string {
-		return spew.Sdump(newPolicy)
-	}))
+func (s *Switch) UpdateForwardingPolicies(
+	chanPolicies map[wire.OutPoint]*channeldb.ChannelEdgePolicy) error {
```
Comment (wpaulino) [inline htlcswitch/switch.go]: This method no longer needs to return an error.

File: htlcswitch/switch.go [hunk 8]
```diff
@@ -445,55 +445,63 @@ func (s *Switch) SendHTLC(firstHop lnwire.ShortChannelID, paymentID uint64,
...
-			}
+	for targetLink := range chanPolicies {
+		cid := lnwire.NewChanIDFromOutPoint(&targetLink)
 
-			linksToUpdate = append(linksToUpdate, link)
+		// If we can't locate a link by its converted channel
+		// ID, then we'll log an error.
+		link, ok := s.linkIndex[cid]
+		if !ok {
+			log.Errorf("unable to find ChannelPoint(%v) to update "+
```
Comment (wpaulino) [inline htlcswitch/switch.go]: Can just log as debug as it's not a fatal error since the link will have the right policy once it becomes active again. Log should also start capitalized.

File: htlcswitch/switch.go [hunk 9]
```diff
@@ -445,55 +445,63 @@ func (s *Switch) SendHTLC(firstHop lnwire.ShortChannelID, paymentID uint64,
...
 	// With all the links we need to update collected, we can release the
 	// mutex then update each link directly.
 	for _, link := range linksToUpdate {
+		newPolicy := dbPolicyToFwdingPolicy(
+			chanPolicies[*link.ChannelPoint()],
+		)
+
+		log.Debugf("Updating link policies: %v",
+			newLogClosure(func() string {
+				return spew.Sdump(newPolicy)
```
Comment (wpaulino) [inline htlcswitch/switch.go]: Any logs containing `spew`s should be trace.

Comment (wpaulino) [reply htlcswitch/switch.go]: This is also going to be logged for every link in its current placement.

File: htlcswitch/link_test.go (hunk 5 above)
Comment (valentinewallace) [reply htlcswitch/link_test.go]: Otherwise, it'll set the max HTLC to zero and generating hops will fail.

File: htlcswitch/switch.go (hunk 7 above)
Comment (valentinewallace) [reply htlcswitch/switch.go]: thanks, fixed

File: htlcswitch/link.go (hunk 4 above)
Comment (valentinewallace) [reply htlcswitch/link.go]: nice lol, fixed

File: rpcserver.go [hunk 10]
```diff
@@ -4612,30 +4612,16 @@ func (r *rpcServer) UpdateChannelPolicy(ctx context.Context,
...
-	err = r.server.htlcSwitch.UpdateForwardingPolicies(p, targetChans...)
-	if err != nil {
-		// If we're unable update the fees due to the links not being
-		// online, then we don't need to fail the call. We'll simply
-		// log the failure.
-		rpcsLog.Warnf("Unable to update link fees: %v", err)
-	}
+	// Finally, we'll apply the set of channel policies to the target
+	// channels' links.
+	r.server.htlcSwitch.UpdateForwardingPolicies(chanPolicies)
```
[the remark this answers is not in the document]
Comment (wpaulino) [reply rpcserver.go]: I'm not opposed to it, but going this route seems like a much bigger change for a bug that's causing routing failures at the moment. This change is pretty small in that it's just exposing the state the gossiper is already acting on. I understand that doing so will just make the separation of concerns worse though.

File: discovery/gossiper.go [hunk 11]
```diff
@@ -1309,9 +1322,13 @@ func (d *AuthenticatedGossiper) processChanPolicyUpdate(
 			edgeInfo.info, edgeInfo.edge,
 		)
 		if err != nil {
-			return nil, err
+			return nil, nil, err
```
[the remark this answers is not in the document]
Comment (wpaulino) [reply discovery/gossiper.go]: This seems unrelated to this PR? In any case, I believe the check is there to automatically upgrade channels to include the field. I don't think we can remove it as we can't ensure all older nodes have gone through this flow.

[the remark this answers is not in the document]
Comment (joostjager) [reply discovery/gossiper.go]: Ok, yes I was thinking of the follow up of this pr.

File: rpcserver.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (joostjager) [reply rpcserver.go]: Ok. I also realize now that links are not always up, so there must be an independent data store somewhere that stores the local policy. 

The only thing I am still thinking about: if we'd just make all policy parameters mandatory for the rpc call (including minhtlc and maxhtlc), would this feedback path of final policies from gossiper still be necessary? If the policy is complete, the links (the ones that are up) and gossiper can be updated independently.

[the remark this answers is not in the document]
Comment (valentinewallace) [reply rpcserver.go]: That's true, but I think it would be best if none of them were mandatory, meaning a user could only specify the ones they want to be updated. That makes me want to make max htlc non-mandatory and lay groundwork for making the other fields non-mandatory as well.

File: htlcswitch/switch.go [hunk 12]
```diff
@@ -445,53 +445,58 @@ func (s *Switch) SendHTLC(firstHop lnwire.ShortChannelID, paymentID uint64,
...
+		cid := lnwire.NewChanIDFromOutPoint(&targetLink)
 
-				return fmt.Errorf("unable to find "+
-					"ChannelPoint(%v) to update link "+
-					"policy", targetLink)
-			}
-
-			linksToUpdate = append(linksToUpdate, link)
+		// If we can't locate a link by its converted channel
+		// ID, then we'll log an error.
```
Comment (wpaulino) [inline htlcswitch/switch.go]: Nit: not an error 😛

File: htlcswitch/switch.go [hunk 13]
```diff
@@ -445,53 +445,58 @@ func (s *Switch) SendHTLC(firstHop lnwire.ShortChannelID, paymentID uint64,
...
-		// policies for the set of targeted links.
-		for _, targetLink := range targetChans {
-			cid := lnwire.NewChanIDFromOutPoint(&targetLink)
-
-			// If we can't locate a link by its converted channel
-			// ID, then we'll return an error back to the caller.
-			link, ok := s.linkIndex[cid]
-			if !ok {
-				s.indexMtx.RUnlock()
+	for targetLink := range chanPolicies {
```
Comment (wpaulino) [inline htlcswitch/switch.go]: Nit: comment above this and the godoc for the method are now outdated.

File: discovery/gossiper.go [hunk 14]
```diff
@@ -343,22 +344,27 @@ func New(cfg Config, selfKey *btcec.PublicKey) *AuthenticatedGossiper {
...
+// sub-systems, then it signs and broadcasts new updates to the network. A
+// mapping between outpoints and updated channel policies is returned, which is
+// used to update the forwarding policies of the underlying links.
 func (d *AuthenticatedGossiper) PropagateChanPolicyUpdate(
-	newSchema routing.ChannelPolicy, chanPoints ...wire.OutPoint) error {
+	newSchema routing.ChannelPolicy, chanPoints ...wire.OutPoint) (
+	map[wire.OutPoint]*channeldb.ChannelEdgePolicy, error) {
 
 	errChan := make(chan error, 1)
+	chanPolicyChan := make(chan map[wire.OutPoint]*channeldb.ChannelEdgePolicy, 1)
```
Comment (joostjager) [inline discovery/gossiper.go]: I'd make a struct containing the map and the error and use only a single channel.

File: discovery/gossiper.go [hunk 15]
```diff
@@ -343,22 +344,27 @@ func New(cfg Config, selfKey *btcec.PublicKey) *AuthenticatedGossiper {
...
+		targetChans:  chanPoints,
+		newSchema:    newSchema,
+		chanPolicies: chanPolicyChan,
+		errResp:      errChan,
 	}
 
 	select {
 	case d.chanPolicyUpdates <- policyUpdate:
-		return <-errChan
+		return <-chanPolicyChan, <-errChan
```
Comment (joostjager) [inline discovery/gossiper.go]: Can these chan reads block forever because there is no select on `d.quit`? I think in the error case, nothing is sent on `chanPolicyChan`.

File: discovery/gossiper_test.go [hunk 16]
```diff
@@ -3353,13 +3353,23 @@ func TestPropagateChanPolicyUpdate(t *testing.T) {
...
 		channelsToAnnounce = append(channelsToAnnounce, newChan)
+
+		// channelPoints ensures that each channel policy in the map
+		// returned by PropagateChanPolicyUpdate has a unique key. Since
+		// the map is keyed by wire.OutPoint, we want to ensure that
+		// each channel has a unique channel point.
+		channelPoints = append(
+			channelPoints,
+			ChannelPoint(wire.OutPoint{Index: uint32(i)}),
+		)
```
Comment (joostjager) [inline discovery/gossiper_test.go]: You could also not build the slice and just create the outpoint inside the `channelsToAnnounce` loop below.

File: htlcswitch/switch.go [hunk 17]
```diff
@@ -445,53 +445,58 @@ func (s *Switch) SendHTLC(firstHop lnwire.ShortChannelID, paymentID uint64,
...
+	log.Tracef("Updating link policies: %v", newLogClosure(func() string {
+		return spew.Sdump(chanPolicies)
+	}))
+
 	for _, link := range linksToUpdate {
+		newPolicy := dbPolicyToFwdingPolicy(
+			chanPolicies[*link.ChannelPoint()],
+		)
+
 		link.UpdateForwardingPolicy(newPolicy)
```
Comment (joostjager) [inline htlcswitch/switch.go]: Why not update in the loop above? As we don't error out anymore, it can be done in a single pass?

File: rpcserver.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (joostjager) [reply rpcserver.go]: Discussed offline. Still not convinced about the business case for adding this granularity on the rpc level. Because the `UpdateChannelPolicy` is currently buggy with zero values, I think there is an opportunity to make the fields mandatory now without it being considered as really breaking (it was broken already). But as discussed, leaving further development out of this pr's scope.

File: htlcswitch/switch.go (hunk 17 above)
Comment (valentinewallace) [reply htlcswitch/switch.go]: good one, fixed

File: rpcserver.go [hunk 18]
```diff
@@ -4615,30 +4615,16 @@ func (r *rpcServer) UpdateChannelPolicy(ctx context.Context,
 	// With the scope resolved, we'll now send this to the
 	// AuthenticatedGossiper so it can propagate the new policy for our
 	// target channel(s).
-	err := r.server.authGossiper.PropagateChanPolicyUpdate(
+	chanPolicies, err := r.server.authGossiper.PropagateChanPolicyUpdate(
 		chanPolicy, targetChans...,
 	)
 	if err != nil {
 		return nil, err
 	}
```
Comment (joostjager) [inline rpcserver.go]: What happens if there are concurrent fee update calls, say call A and call B, for the same channel? I think it is possible that A updates the db, then B updates the db, then B applies policy B to the link and finally A applies policy A to the link. At that point the db and link are no longer in sync.

Comment (wpaulino) [reply rpcserver.go]: Looks like a pre-existing issue, but it seems unlikely given that the gossiper can only handle one `PropagateChanPolicyUpdate` call at a time? B would start after A is done, at which point A is likely to have reached the switch already.

Comment (joostjager) [reply rpcserver.go]: Yes, indeed looks unlikely, but it still is a small gap. And pre-existing indeed. Maybe we can just throw a lock around it all? Or at least add a TODO describing this. Or do nothing at all, given that it is pre-existing.

Comment (valentinewallace) [reply rpcserver.go]: Added a lock around the db+link update code!

File: rpcserver.go [hunk 19]
```diff
@@ -4617,30 +4623,21 @@ func (r *rpcServer) UpdateChannelPolicy(ctx context.Context,
 	// With the scope resolved, we'll now send this to the
 	// AuthenticatedGossiper so it can propagate the new policy for our
 	// target channel(s).
-	err := r.server.authGossiper.PropagateChanPolicyUpdate(
+	policyUpdateLock.Lock()
+	chanPolicies, err := r.server.authGossiper.PropagateChanPolicyUpdate(
 		chanPolicy, targetChans...,
 	)
 	if err != nil {
 		return nil, err
```
Comment (joostjager) [inline rpcserver.go]: lock isn't released

Comment (valentinewallace) [reply rpcserver.go]: Released it [here](https://github.com/lightningnetwork/lnd/pull/3439/files#diff-01ada5680aaa18fc79cbf78b2d55a37bR4640), otherwise isn't there a chance the aforementioned concurrency problem could still happen?

Comment (valentinewallace) [reply rpcserver.go]: Released it [here](https://github.com/lightningnetwork/lnd/pull/3439/files#diff-01ada5680aaa18fc79cbf78b2d55a37bR4640), otherwise isn't there a chance the aforementioned concurrency problem could still happen?

Comment (joostjager) [reply rpcserver.go]: I mean in the error case

Comment (valentinewallace) [reply rpcserver.go]: Ahh. Thanks, fixed
